cmd/graph.go

```diff
@@ -1,35 +1,110 @@
 package cmd
 
 import (
+	"encoding/json"
+	"html/template"
+	"net/http"
+	"sort"
+
 	log "github.com/sirupsen/logrus"
 	"github.com/spf13/cobra"
 	"github.com/srl-wim/container-lab/clab"
 )
 
+const (
+	templatePath = "/etc/containerlab/templates/d3js/index.html"
+)
+
+var srv string
+var tmpl string
+
+type graphTopo struct {
+	Nodes []containerDetails `json:"nodes,omitempty"`
+	Links []link             `json:"links,omitempty"`
+}
+type link struct {
+	Source         string `json:"source,omitempty"`
+	SourceEndpoint string `json:"source_endpoint,omitempty"`
+	Target         string `json:"target,omitempty"`
+	TargetEndpoint string `json:"target_endpoint,omitempty"`
+}
+
+type topoData struct {
+	Name string
+	Data template.JS
+}
+
 // graphCmd represents the graph command
 var graphCmd = &cobra.Command{
 	Use:   "graph",
 	Short: "generate a topology graph",
 
-	Run: func(cmd *cobra.Command, args []string) {
-		opts := []clab.ClabOption{
-			clab.WithDebug(debug),
-			clab.WithTimeout(timeout),
-			clab.WithTopoFile(topo),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		c := clab.NewContainerLab(debug)
+
+		if err := c.GetTopology(topo); err != nil {
+			return err
 		}
-		c := clab.NewContainerLab(opts...)
 
 		// Parse topology information
 		if err := c.ParseTopology(); err != nil {
-			log.Fatal(err)
+			return err
+		}
+
+		if srv == "" {
+			if err := c.GenerateGraph(topo); err != nil {
+				return err
+			}
+			return nil
+		}
+		gtopo := graphTopo{
+			Nodes: make([]containerDetails, 0, len(c.Nodes)),
+			Links: make([]link, 0, len(c.Links)),
+		}
+		for name, n := range c.Nodes {
+			gtopo.Nodes = append(gtopo.Nodes, containerDetails{
+				Name:  name,
+				Kind:  n.Kind,
+				Image: n.Image,
+				Group: n.Group,
+			})
+		}
+		sort.Slice(gtopo.Nodes, func(i, j int) bool {
+			return gtopo.Nodes[i].Name < gtopo.Nodes[j].Name
+		})
+		for _, l := range c.Links {
+			gtopo.Links = append(gtopo.Links, link{
+				Source:         l.A.Node.ShortName,
+				SourceEndpoint: l.A.EndpointName,
+				Target:         l.B.Node.ShortName,
+				TargetEndpoint: l.B.EndpointName,
+			})
+		}
+		b, err := json.Marshal(gtopo)
```

Comment:
add a message what is this dump about for a debug reader to understand it better?

Reply:
sorry, wrong line, it was about the debug statement a few lines below

Reply:
done